protocols/raft/src/main/java/io/atomix/protocols/raft/roles/LeaderRole.java

```diff
@@ -682,19 +684,37 @@ private boolean initializing() {
 
     final long sequenceNumber = request.sequenceNumber();
 
+    // If a command with the given sequence number is already pending, return the existing future to ensure
+    // duplicate requests aren't committed as duplicate entries in the log.
+    PendingCommand existingCommand = session.getCommand(sequenceNumber);
+    if (existingCommand != null) {
+      if (sequenceNumber == session.nextRequestSequence()) {
+        session.removeCommand(sequenceNumber);
+        commitCommand(existingCommand.request(), existingCommand.future());
+        session.setRequestSequence(sequenceNumber);
+        drainCommands(session);
+      }
+      return existingCommand.future();
+    }
+
     final CompletableFuture<CommandResponse> future = new CompletableFuture<>();
 
-    // If the session's current sequence number is less then one prior to the request sequence number, reject
-    // the command to force it to be resent in the correct order. Note that it's possible for the session
-    // sequence number to be greater than the request sequence number. In that case, it's likely that the
-    // command was submitted more than once to the cluster, and the command will be deduplicated once
-    // applied to the state machine.
+    // If the request sequence number is greater than the next sequence number, that indicates a command is missing.
```

Comment:
Clients’ commands are appended to the Raft log in sequential order. But when clients’ requests are delivered out of order e.g. after leader changes, this can often lead to significantly increased communication between the leader and client. Queueing command’s to be written to the Raft log significantly decreases that communication, reduces latency, and thus reduces the frequency of timeouts for blocking commands.